src/transaction/log_checkpoint_info.cpp

```diff
@@ -286,5 +287,164 @@ namespace cublog
+	    return;
+	  }
+	chkpt_one = &chkpt;
+
+	/*
+	 * Clear the transaction since it may have old stuff in it.
+	 * Use the one that is find in the checkpoint record
+	 */
+	logtb_clear_tdes (thread_p, tdes);
+
+	tdes->isloose_end = chkpt_one->isloose_end;
+	if (chkpt_one->state == TRAN_ACTIVE || chkpt_one->state == TRAN_UNACTIVE_ABORTED)
+	  {
+	    tdes->state = TRAN_UNACTIVE_UNILATERALLY_ABORTED;
+	  }
+	else
+	  {
+	    tdes->state = chkpt_one->state;
+	  }
+	LSA_COPY (&tdes->head_lsa, &chkpt_one->head_lsa);
+	LSA_COPY (&tdes->tail_lsa, &chkpt_one->tail_lsa);
+	LSA_COPY (&tdes->undo_nxlsa, &chkpt_one->undo_nxlsa);
+	LSA_COPY (&tdes->posp_nxlsa, &chkpt_one->posp_nxlsa);
+	LSA_COPY (&tdes->savept_lsa, &chkpt_one->savept_lsa);
+	LSA_COPY (&tdes->tail_topresult_lsa, &chkpt_one->tail_topresult_lsa);
+	LSA_COPY (&tdes->rcv.tran_start_postpone_lsa, &chkpt_one->start_postpone_lsa);
+	tdes->client.set_system_internal_with_user (chkpt_one->user_name);
+	if (LOG_ISTRAN_2PC (tdes))
+	  {
+	    m_has_2pc = true;
+	  }
+      }
+
+
+    if (area != NULL)
+      {
+	free_and_init (area);
+      }
+
+    /*
+     * Now add top system operations that were in the process of
+     * commit to this transactions
+     */
+
+    log_page_local = (LOG_PAGE *) PTR_ALIGN (log_page_buffer, MAX_ALIGNMENT);
+    log_page_local->hdr.logical_pageid = NULL_PAGEID;
+    log_page_local->hdr.offset = NULL_OFFSET;
+
+    for (auto sysop : m_sysops)
+      {
+	chkpt_topone = &sysop;
+	tdes = logtb_rv_find_allocate_tran_index (thread_p, chkpt_topone->trid, &NULL_LSA);
+	if (tdes == NULL)
+	  {
+	    if (area != NULL)
+	      {
+		free_and_init (area);
+	      }
+
+	    logpb_fatal_error (thread_p, true, ARG_FILE_LINE, "log_recovery_analysis");
+	    return;
+	  }
+
+	if (tdes->topops.max == 0 || (tdes->topops.last + 1) >= tdes->topops.max)
+	  {
+	    if (logtb_realloc_topops_stack (tdes, m_sysops.size()) == NULL)
+	      {
+		if (area != NULL)
+		  {
+		    free_and_init (area);
+		  }
+
+		logpb_fatal_error (thread_p, true, ARG_FILE_LINE, "log_recovery_analysis");
+		return;
+	      }
+	  }
+
+	if (tdes->topops.last == -1)
+	  {
+	    tdes->topops.last++;
+	  }
+	else
+	  {
+	    assert (tdes->topops.last == 0);
+	  }
+	tdes->rcv.sysop_start_postpone_lsa = chkpt_topone->sysop_start_postpone_lsa;
+	tdes->rcv.atomic_sysop_start_lsa = chkpt_topone->atomic_sysop_start_lsa;
+	log_lsa_local = chkpt_topone->sysop_start_postpone_lsa;
+	error_code =
+		log_read_sysop_start_postpone (thread_p, &log_lsa_local, log_page_local, false, &sysop_start_postpone,
+					       NULL, NULL, NULL, NULL);
+	if (error_code != NO_ERROR)
+	  {
+	    assert (false);
+	    return;
+	  }
+	tdes->topops.stack[tdes->topops.last].lastparent_lsa = sysop_start_postpone.sysop_end.lastparent_lsa;
+	tdes->topops.stack[tdes->topops.last].posp_lsa = sysop_start_postpone.posp_lsa;
+      }
+  }
+  void
+  checkpoint_info::recovery_2pc_analysis (THREAD_ENTRY *thread_p) const
+  {
+    if (m_has_2pc)
```

Comment:
Early out if there is no 2pc.
```suggestion
    if (!m_has_2pc)
```